tests/S3/ApplyChecksumMiddlewareTest.php

```diff
@@ -61,4 +61,50 @@ public function getContentMd5UseCases()
             ],
         ];
     }
+
+    /**
+     * @dataProvider getContentSha256UseCases
+     */
+    public function testAddsContentSHA256AsAppropriate($operation, $args, $hashAdded, $hashValue)
+    {
+        $s3 = $this->getTestClient('s3');
+        $this->addMockResults($s3, [[]]);
+        $command = $s3->getCommand($operation, $args);
+        $command->getHandlerList()->appendBuild(
+            Middleware::tap(function ($cmd, RequestInterface $request) use ($hashAdded, $hashValue) {
+                $this->assertSame($hashAdded, $request->hasHeader('x-amz-content-sha256'));
+                $this->assertEquals($hashValue, $request->getHeaderLine('x-amz-content-sha256'));
+            })
+        );
+        $s3->execute($command);
+    }
+
+    public function getContentSha256UseCases()
+    {
+        $hash = 'SHA256HASH';
+
+        return [
+            // Do nothing if ContentSHA256 was not provided.
```

Comment:
@jeremeamia The `S3SignatureV4` will always try to set the `x-amz-content-sha256` header to the hash of the request body or POST fields on a `PutObject`.